src/cardsort/analysis.py

```diff
@@ -1,30 +1,35 @@
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+from scipy.cluster import hierarchy
+from scipy.spatial.distance import squareform
+from typing import List, Union
+
 __all__ = [
     "get_distance_matrix",
     "create_dendrogram",
     "get_cluster_labels",
     "get_cluster_labels_df",
 ]
 
-import numpy as np
-import pandas as pd
-import matplotlib.pyplot as plt
-from scipy.cluster import hierarchy
-from scipy.spatial.distance import squareform
-
 
-def _get_distance_matrix_for_user(df_user):
+def _get_distance_matrix_for_user(df_user: pd.DataFrame) -> np.ndarray:
     n = len(df_user)
 
     X = np.zeros((n, n))
     for i in range(n):
         for j in range(n):
```

Comment:
You would save a bunch of computations if you use here:

```python
for i in range(n):
    for j in range(i, n):
```
so you only iterate over the upper triangle of your symmetric matrix. You are already copying the entries to the lower triangle below:

```python
X[i, j] = X[j, i] = 0 if cat1 == cat2 else 1
```

Reply:
Thank you, fixed it